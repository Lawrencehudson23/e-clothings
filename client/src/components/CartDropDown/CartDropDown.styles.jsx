import styled from "styled-components";
import CustomButton from "../CustomButton/CustomButton";

export const CartDropDownContainer = styled.div`
  position: absolute;
  width: 240px;
  height: 340px;
  display: flex;
  flex-direction: column;
  padding: 20px;
  border: 1px solid black;
  background-color: white;
  top: 90px;
  right: 40px;
  z-index: 5;
`;
export const CartDropDownButton = styled(CustomButton)`
  margin-top: auto;
`;
CartDropDownButton.displayname = "CartDropDownButton";

export const EmptyMessageContainer = styled.span`
  font-size: 18px;
  margin: 50px auto;
`;

EmptyMessageContainer.displayname = "EmptyMessageContainer";

export const CartItemsContainer = styled.div`
  height: 240px;
  display: flex;
  flex-direction: column;
  overflow: scroll;
`;
