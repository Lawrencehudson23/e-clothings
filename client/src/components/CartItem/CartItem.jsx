import React from "react";

import "./CartItem.scss";
const CartItem = ({ item }) => {
  const { imageUrl, name, price, quantity } = item;
  return (
    <div className="cart-item">
      <img src={imageUrl} alt="item" />
      <div className="item-details">
        <span className="name">{name}</span>
        <span className="price">
          {quantity} X <span className="dollar">$</span>
          {price}
        </span>
      </div>
    </div>
  );
};

export default CartItem;
