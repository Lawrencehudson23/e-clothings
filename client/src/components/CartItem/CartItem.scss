.cart-item {
    width: 100%;
    display: flex;
    height: 80px;
    margin-bottom: 15px;

    img {
        width: 30%;
    }

    .item-details {
        width: 70%;
        display: flex;
        flex-direction: column;
        align-items: flex-start;
        justify-content: center;
        padding: 10px 20px;
        font-size: 15px;

        .dollar {
            font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen, Ubuntu, Cantarell, 'Open Sans', 'Helvetica Neue', sans-serif;
            font-weight: 700;
        }
    }
}