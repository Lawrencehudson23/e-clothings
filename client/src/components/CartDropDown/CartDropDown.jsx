import React from "react";
import { connect } from "react-redux";
import { createStructuredSelector } from "reselect";
import CartItem from "../CartItem/CartItem";
import { withRouter } from "react-router-dom";
import { selectCartItems } from "../../redux/cart/cart.selectors";
import { toggleCartHidden } from "../../redux/cart/cart.actions";

import {
  CartDropDownContainer,
  CartItemsContainer,
  EmptyMessageContainer,
  CartDropDownButton,
} from "./CartDropDown.styles";

export const CartDropDown = ({ cartItems, history, dispatch }) => (
  <CartDropDownContainer>
    <CartItemsContainer>
      {cartItems.length ? (
        cartItems.map((cartItem) => (
          <CartItem key={cartItem.id} item={cartItem} />
        ))
      ) : (
        <EmptyMessageContainer>Your cart is empty</EmptyMessageContainer>
      )}
    </CartItemsContainer>
    <CartDropDownButton
      onClick={() => {
        history.push("/checkout");
        dispatch(toggleCartHidden());
      }}
    >
      GO TO CHECKOUT
    </CartDropDownButton>
  </CartDropDownContainer>
);

const mapStateToProps = createStructuredSelector({
  cartItems: selectCartItems,
});
//NOTE if don't supply mapDispatchToProps as the second parameter. connect will pass the dispatch as a property
export default withRouter(connect(mapStateToProps)(CartDropDown));
