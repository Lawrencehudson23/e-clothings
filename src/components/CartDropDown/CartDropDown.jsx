import React from "react";
import { connect } from "react-redux";
import { createStructuredSelector } from "reselect";
import CustomButton from "../CustomButton/CustomButton";
import CartItem from "../CartItem/CartItem";
import { withRouter } from "react-router-dom";
import "./CartDropDown.scss";
import { selectCartItems } from "../../redux/cart/cart.selectors";
import { toggleCartHidden } from "../../redux/cart/cart.actions";
const CartDropDown = ({ cartItems, history, dispatch }) => (
  <div className="cart-dropdown">
    <div className="cart-items">
      {cartItems.length ? (
        cartItems.map((cartItem) => (
          <CartItem key={cartItem.id} item={cartItem} />
        ))
      ) : (
        <span className="empty-message">Your cart is empty</span>
      )}
    </div>
    <CustomButton
      onClick={() => {
        history.push("/checkout");
        dispatch(toggleCartHidden());
      }}
    >
      GO TO CHECKOUT
    </CustomButton>
  </div>
);

const mapStateToProps = createStructuredSelector({
  cartItems: selectCartItems,
});
//NOTE if don't supply mapDispatchToProps as the second parameter. connect will pass the dispatch as a property
export default withRouter(connect(mapStateToProps)(CartDropDown));
