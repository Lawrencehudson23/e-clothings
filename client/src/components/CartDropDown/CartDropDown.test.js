import React from "react";
import { shallow } from "enzyme";
import { CartDropDown } from "./CartDropDown";
import CartItem from "../CartItem/CartItem";
describe("CartDropDown component", () => {
  let wrapper;
  let mockHistory;
  let mockDispatch;
  const mockCartItems = [{ id: 1 }, { id: 2 }, { id: 3 }];

  beforeEach(() => {
    mockHistory = {
      push: jest.fn(),
    };

    mockDispatch = jest.fn();

    const mockProps = {
      cartItems: mockCartItems,
      history: mockHistory,
      dispatch: mockDispatch,
    };

    wrapper = shallow(<CartDropDown {...mockProps} />);
  });

  it("should render CartDropDown component", () => {
    expect(wrapper.debug()).toMatchSnapshot();
  });
  // TODO: button in another level?
  it.skip("should call history.push when button is clicked", () => {
    wrapper.find("CartDropDownButton").simulate("click");
    expect(mockHistory.push).toHaveBeenCalled();
    expect(mockDispatch).toHaveBeenCalledWith(toggleCartHidden());
  });

  it("should render an equal number of CartItem components as the cartItems prop", () => {
    expect(wrapper.find(CartItem).length).toEqual(mockCartItems.length);
  });

  it("should render EmptyMessageContainer if CartItems is empty", () => {
    const mockProps = {
      cartItems: [],
      history: mockHistory,
      dispatch: mockDispatch,
    };

    const newWrapper = shallow(<CartDropDown {...mockProps} />);
    //FIXME: should be true?
    expect(newWrapper.exists("EmptyMessageContainer")).toBe(false);
  });
});
