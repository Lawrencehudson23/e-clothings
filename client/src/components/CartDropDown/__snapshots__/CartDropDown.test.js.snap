// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartDropDown component should render CartDropDown component 1`] = `
"<styled.div>
  <styled.div>
    <CartItem item={{...}} />
    <CartItem item={{...}} />
    <CartItem item={{...}} />
  </styled.div>
  <Styled(CustomButton) onClick={[Function: onClick]}>
    GO TO CHECKOUT
  </Styled(CustomButton)>
</styled.div>"
`;
